pyro/contrib/forecast/forecaster.py

```diff
@@ -283,3 +284,64 @@ def forward(self, data, covariates, num_samples):
             with PrefixConditionMessenger(self.model._prefix_condition_data):
                 with pyro.plate("particles", num_samples, dim=dim):
                     return self.model(data, covariates)
+
+
+class HMCForecaster(nn.Module):
+    """
+    Forecaster for a :class:`ForecastingModel` using Hamiltonian Monte Carlo.
+
+    On initialization, this will run MCMC to get posterior samples of the model.
+
+    After construction, this can be called to generate sample forecasts.
+
+    :param ForecastingModel model: A forecasting model subclass instance.
+    :param data: A tensor dataset with time dimension -2.
+    :type data: ~torch.Tensor
+    :param covariates: A tensor of covariates with time dimension -2.
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+
+    :param int num_warmup: number of MCMC warmup steps.
+    :param int num_samples: number of MCMC samples.
+    :param int num_chains: number of parallel MCMC chains.
+    """
+    def __init__(self, model, data, covariates=None, *,
+                 num_warmup=1000, num_samples=1000, num_chains=1):
+        assert data.size(-2) == covariates.size(-2)
+        super().__init__()
+        self.model = model
+        self.num_samples = num_samples * num_chains
+        max_plate_nesting = _guess_max_plate_nesting(model, (data, covariates), {})
+        self.max_plate_nesting = max(max_plate_nesting, 1)  # force a time plate
+
+        mcmc = MCMC(NUTS(model), warmup_steps=num_warmup, num_samples=num_samples, num_chains=num_chains)
+        mcmc.run(data, covariates)
+        mcmc.summary()
+
+        # inspect the model with particles plate = 1, so that we can reshape samples to
+        # add any missing plate dim in front.
+        with poutine.trace() as tr:
+            with pyro.plate("particles", 1, dim=-self.max_plate_nesting - 1):
+                model(data, covariates)
+
+        self._trace = tr.trace
+        samples = mcmc.get_samples()
+        for name, node in list(self._trace.nodes.items()):
+            if name not in samples:
+                del self._trace.nodes[name]
+            else:
+                sample = samples[name]
+                node['value'] = sample.reshape(
+                    (self.num_samples,) + (1,) * (node['value'].dim() - sample.dim()) + sample.shape[1:])
+
+    @torch.no_grad()
+    def forward(self, data, covariates):
```

Comment:
Do we need to support `num_samples` (by subsampling from mcmc samples) here? I recall that @martinjankowiak does not like that idea because MCMC samples are expensive.

Reply:
I would like to keep the interfaces identical, so yes I'd like to keep a `num_samples` kwarg. This is also required for `crps` evaluation and plotting quantiles.

Can we not simply reuse the existing samples?

Reply:
So I'll support that.

> Can we not simply reuse the existing samples?

In other frameworks, they use subsampling (Stan, Rlgt, PyMC3).

Reply:
@fritzo I just support `num_samples` with subsampling.